module_3/gifts/service/src/main/java/com/epam/esm/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,338 @@
+		errors = orderValidation.validateOrderCertificatesAmountRequirements(orderCertificates);
+		if (!errors.isEmpty()) {
+			throw new ValidationException(errors, ErrorCode.INVALID_CERTIFICATE);
+		}
+
+		BigDecimal cost = new BigDecimal("0");
+
+		for (OrderCertificateDto orderCertificate : orderCertificates) {
+			Long certificateId = orderCertificate.getCertificate().getId();
+			Optional<CertificateModel> certificateModelOptional = certificateRepository.findById(certificateId);
+			CertificateModel certificateModel = certificateModelOptional.get();
+
+			int certificateAmount = orderCertificate.getCertificateAmount();
+
+			cost = cost.add(certificateModel.getPrice().multiply(new BigDecimal(certificateAmount)));
+		}
+
+		orderDto.setId(null);
+		orderDto.setCertificates(orderCertificates);
+		OrderModel orderToSave = orderConverter.convertToModel(orderDto);
+		orderToSave.setCost(cost);
+		orderToSave.setUser(userModel.get());
+		return orderToSave;
+	}
+
+	/**
+	 * Reads information about the order with passed id for the specified user.
+	 * 
+	 * @param userId  id of the user whose order should be read
+	 * @param orderId id of the order to be read
+	 * @return information about the order with passed id for the specified user
+	 * @throws NotFoundException   if the order with passed id does not exist
+	 * @throws ValidationException if passed order or user id is invalid
+	 */
+	@Override
+	public OrderDataDto readOrderDataByUserId(long userId, long orderId) {
+		if (!ValidationUtil.isPositive(orderId)) {
+			throw new ValidationException(EntityConstant.ID + ValidationUtil.ERROR_RESOURCE_DELIMITER + orderId,
+					ErrorCode.INVALID_ORDER_ID);
+		}
+
+		Optional<OrderModel> orderModel = orderRepository.findById(orderId);
+
+		if (orderModel.isEmpty()) {
+			throw new NotFoundException(EntityConstant.ID + ValidationUtil.ERROR_RESOURCE_DELIMITER + orderId,
+					ErrorCode.NO_ORDER_FOUND);
+		} else if (orderModel.get().getUser().getId() != userId) {
+			StringBuilder errorMessage = new StringBuilder();
+			errorMessage.append(EntityConstant.USER_ID + ValidationUtil.ERROR_RESOURCE_DELIMITER + userId);
+			errorMessage.append(ValidationUtil.ERROR_RESOURCES_LIST_DELIMITER);
+			errorMessage.append(EntityConstant.ORDER_ID + ValidationUtil.ERROR_RESOURCE_DELIMITER + orderId);
+			throw new ValidationException(errorMessage.toString(), ErrorCode.USER_ID_MISMATCH);
+		}
+
+		OrderDataDto orderDataDto = orderDataConverter.convertToDto(orderModel.get());
+
+		return orderDataDto;
+	}
+
+	/**
+	 * Reads all orders according to the passed parameters.
+	 * 
+	 * @param params the parameters which define the choice of orders and their
+	 *               ordering
+	 * @return orders which meet passed parameters
+	 * @throws ValidationException if passed parameters are invalid
+	 */
+	@Override
+	public List<OrderDto> readAll(MultiValueMap<String, String> params) {
```

Comment:
as I see you try to implement pagination then use should return dto that contains the number of all page, the current page etc. and dto should extend PagedModel

Reply:
@maximBortnik, There wasn't mentioned in the task that pagination includes returning the number of pages and current page, but I understand that it is convenient for usage (ex., for UI). I will add this information in the returned dto in future as I will not manage to implement fully tested solution till the tomorrow's demo.

Reply:
@LizavetaSinitsyna In fact, it will never be mentioned. It's like a standard approach to pagination implementation.